dist/js/common-header.js

```
@@ -3928,6 +3901,117 @@ angular.module("risevision.common.geodata", [])
 
 })(angular);
 
+(function (angular){
+  
+  "use strict";
+  
+  angular.module("risevision.common.companystate", ["risevision.core.company"])
+  
+  .factory("companyState", ["$location", "getCompany", "objectHelper", 
+    "$rootScope", "$log", "$q",
+  function ($location, getCompany, objectHelper, $rootScope, $log, $q) {
+    var pendingSelectedCompany;
+    
+    var _state = {
+      userCompany: {},
+      selectedCompany: {}
+    };
+    
+    var _resetCompanyState = function () {
+      objectHelper.clearObj(_state.selectedCompany);
+      objectHelper.clearObj(_state.userCompany);
+      $log.debug("Company state has been reset.");
+    };
+    
+    if($location.search().cid) {
+      $log.debug("cid", $location.search().cid, "saved for later processing.");
+      pendingSelectedCompany = $location.search().cid;
+    }
+    
+    var _init = function () {
+      var deferred = $q.defer();
+      
+      //populate userCompany
+      getCompany().then(function(company) {
+        objectHelper.clearAndCopy(company, _state.userCompany);
+                
+        return _switchCompany(pendingSelectedCompany);
```

@olegrise I agree this code is not very clear. 

In `_switchCompany` I check the `companyId` and call `resetCompany` if it is blank. Also, if the `getCompany` call fails, the promise will fail and the `resetCompany` will be called in `_init`.

However, my logic is the following: 
If the user is in their sub-company, and adds a cid that fails to produce a valid company, I didn't want to `resetCompany` on them in that case. The call will just fail silently (updated console log) and not change the state.

Respectively, if switchCompany is called with a blank companyId, it will just call `resetCompany`.